clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -1034,28 +1024,35 @@ private Record nextFetchedRecord() {
                     records = currentBatch.streamingIterator();
                 }
 
-                Record record = records.next();
-                maybeEnsureValid(record);
-
-                // skip any records out of range
-                if (record.offset() >= nextFetchOffset) {
-                    nextFetchOffset = record.offset() + 1;
-
-                    // control records are not returned to the user
-                    if (!currentBatch.isControlBatch())
-                         return record;
+                // If an exception was thrown from the last record. We should throw the same exception here again.
+                // We do not cache the exception thrown last time because the stack trace could be different
+                // from last time.
+                if (hasExceptionInLastFetch) {
+                    maybeEnsureValid(lastRecord);
```

Comment:
One of the possibilities for a corrupt record is an error transmitting over the network (the TCP checksum is only 2 bytes). We could recover from this error by discarding the current batch and refetching from the failed offset. The downside is, well, we have to refetch. In practice, I assume this would be super rare, but maybe it's still worth allowing for the possibility? For parsing errors, refetching may not help, so caching the error seems fair.